### PR TITLE
Solution For Longest Consecutive Sequence Challenge

### LINKED ISSUE #19: Checking the longest consecutive sequence in a list
## Description

Write a function that checks the longest consecutive sequence in a list.

---

## Behavior

The function should:
- Take a list of integers.
- Returns the number of longest consecutive sequences of the integers in the list.
- Give error when the elements in the list is not of type integer such as: none, nan and string.

---

### Example 1

    Input:
    function([1,4,5,2,])

    Output:
    2

### Example 2

    Input:
    function([3,6,4,7,5,10])

    Output:
    5

### Example 3

    Input:
    function([1,2,none, 6,10])

    Output:
    Error: "The list should only contain integers"

### Helpful Links

Here is the [link](https://leetcode.com/problems/longest-consecutive-sequence/) to the challenge on the LeetCode website.

### PR DESCRIPTION
**Pull request by hamid4231**
---
Longest Consecutive Sequence Challenge
Code review checklist for the Python challenge
---

## Behavior

The function should:
- Take a list of integers.
- Returns the number of longest consecutive sequences of the integers in the list.
- Give error when the elements in the list is not of type integer such as: none, nan and string.

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/longest_consecutive_sequence.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    Args:
+        nums (List[int]): A list of integers.
+
+    Returns:
+        int: The length of the longest consecutive sequence.
+
+    Raises:
+        TypeError: If `nums` is not a list.
+        TypeError: If `nums` contains non-integer elements.
```
Comment (KarimMakki) [inline solutions/longest_consecutive_sequence.py]: You documented that it raises a TypeError but your code raises AssertionError instead, please modify this to be AssertionError rather than TypeError for both of them

File: solutions/tests/test_longest_consecutive_sequence.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+This module provides unit tests for the function calculate_discount.
+Created on: 2025/1/5
+Author: Hamidullah Rajabi
+"""
```
Comment (KarimMakki) [inline solutions/tests/test_longest_consecutive_sequence.py]: The test module's docstring should include a description of the test categories you've implemented. Currently, you have organized your tests into three clear categories:

1. Regular Passing Cases
2. Rare or Edge Cases
3. Defensive Cases That Raise Errors 

**Consider adding these categories to your module docstring like this:**

```
"""
This module provides unit tests for the longest_consecutive_sequence function.

Test Categories:
Regular Passing Cases - Tests with typical inputs
Rare or Edge Cases - Tests with empty lists, single elements, and duplicates
Defensive Cases - Tests that verify error handling for invalid inputs

Created on: 2025/1/5
Author: Hamidullah Rajabi
"""
```

File: solutions/longest_consecutive_sequence.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        ValueError: If the list contains `None`values.
+        ValueError: If the list contains `NaN` values.
+
+    Doctests:
```
Comment (KarimMakki) [inline solutions/longest_consecutive_sequence.py]: According to Python's docstring conventions, it should be "Examples" rather than "Doctests".

File: solutions/longest_consecutive_sequence.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            current_num = num
+            current_length = 1
+
+            while current_num + 1 in nums_set:
+                current_num += 1
+                current_length += 1
+
+            max_length = max(max_length, current_length)
+
+    return max_length
```
Comment (KarimMakki) [inline solutions/longest_consecutive_sequence.py]: Your implementation looks good, but consider adding clear comments describing the different stages of the strategy

**Review by KarimMakki — CHANGES_REQUESTED** [review]
Hello @hamid4231 , I've made few comments regarding some modifications to your code and the docstrings, please check them.

File: solutions/longest_consecutive_sequence.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+the longest consecutive sequence in a list of integers.
+Created on: 2025/1/5
+@author: Hamidullah Rajabi
+"""
+
+"""
+longest_consecutive_sequence.py
+
+This module provides a function to find the length of the longest consecutive sequence in a list of integers.
+"""
```
Comment (KarimMakki) [inline solutions/longest_consecutive_sequence.py]: This is actually a duplicate of the module docstring, please remove this one and keep the above.

**Comment by hamid4231** [comment]
> Hello @hamid4231 , I've made few comments regarding some modifications to your code and the docstrings, please check them.

Hello @KarimMakki, thanks for the constructive feedbacks on my solution. I have adjusted my code according to your suggestions. Please feel free to check and let me know if it needs further adjustments.

File: solutions/longest_consecutive_sequence.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        nums (List[int]): A list of integers.
+
+    Returns:
+        int: The length of the longest consecutive sequence.
+
+    Raises:
+        AssertionError: If `nums` is not a list.
+        AssertionError: If `nums` contains non-integer elements.
+        ValueError: If the list contains `None`values.
+        ValueError: If the list contains `NaN` values.
```
Comment (KarimMakki) [inline solutions/longest_consecutive_sequence.py]: I believe you may make it looks more organized like this:

 
```
Raises:
        AssertionError:
        -  If `nums` is not a list.
        -  If `nums` contains non-integer elements
        ValueError:
        -  If the list contains `None`values
        -  If the list contains `NaN` values
```

**Review by KarimMakki — APPROVED** [review]
Thank you @hamid4231 for doing all the changes! your code looks good now.
